### PR TITLE
Fix url_extract_* functions for malformed URL's

### LINKED ISSUE #7038: url_extract_path should return NULL if URI is malformed
### Bug description

In Presto, url_extract_path returns NULL if URL is malformed, url_decode throws.

Velox currently throws in both functions. Need to fix to match Presto.

```
presto:di>  select url_extract_path('https://www.ucu.edu.uy/agenda/evento/%%UCUrlCompartir%%');
 _col0
-------
 NULL
(1 row)

presto:di> select url_decode('https://www.ucu.edu.uy/agenda/evento/%%UCUrlCompartir%%');
Query 20231013_105704_33253_divi2 failed: URLDecoder: Illegal hex characters in escape (%) pattern - Error at index 0 in: "%U"
```

CC: @zacw7 @amitkdutta @aditi-pandit @kgpai 

### System information

n/a

### Relevant logs

_No response_

### PR DESCRIPTION
**Pull request by kgpai**
Fixes #7038 

Malformed URL's which contain invalid escape sequences (%xx) used to throw in Velox, but not in Presto.

Also, for absolute URI's, url_extract_path used to return NULL when it should return the path, e.g url_extract_path('foo') should return 'foo'. Fix this by making the scheme/authority/path regex to be compliant with the URI RFC  (https://www.rfc-editor.org/rfc/rfc3986#appendix-A).
 
Some examples of the new changes:

``` 
> SELECT url_extract_path('https://www.ucu.edu.uy/agenda/evento/%%UCUrlCompartir%%'); 

Before: throws exception.
After: returns NULL.

> SELECT url_extract_path('foo');

Before: returns NULL.
After: returns 'foo'.
```

### CODE REVIEW

File: velox/functions/prestosql/URLFunctions.h [hunk 1]
```diff
@@ -30,10 +31,10 @@ FOLLY_ALWAYS_INLINE StringView submatch(const boost::cmatch& match, int idx) {
 template <typename TInString>
 bool parse(const TInString& rawUrl, boost::cmatch& match) {
   static const boost::regex kUriRegex(
-      "([a-zA-Z][a-zA-Z0-9+.-]*):" // scheme:
-      "([^?#]*)" // authority and path
-      "(?:\\?([^#]*))?" // ?query
-      "(?:#(.*))?"); // #fragment
+      "^(([^:\\/?#]+):)?" // scheme:
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:34]: These regexes are not very readable. Maybe add comments to explain each of them?

File: velox/functions/prestosql/URLFunctions.h [hunk 2]
```diff
@@ -84,6 +85,35 @@ FOLLY_ALWAYS_INLINE void urlEscape(TOutString& output, const TInString& input) {
   output.resize(outIndex);
 }
 
+template <typename TInString>
+FOLLY_ALWAYS_INLINE bool urlUnescapeCheck(const TInString& input) {
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:89]: Would you add a comment to explain what checks are performed here?

naming: start with a verb (checkUrlUnescape) or isValidUrl

File: velox/functions/prestosql/URLFunctions.h [hunk 3]
```diff
@@ -84,6 +85,35 @@ FOLLY_ALWAYS_INLINE void urlEscape(TOutString& output, const TInString& input) {
   output.resize(outIndex);
 }
 
+template <typename TInString>
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:88]: Is this template parameter needed? Is not not always a StringView?

File: velox/functions/prestosql/URLFunctions.h (hunk 2 above)
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:89]: Remove FOLLY_ALWAYS_INLINE. This function is too long.

File: velox/functions/prestosql/URLFunctions.h [hunk 4]
```diff
@@ -143,11 +173,16 @@ struct UrlExtractProtocolFunction {
   FOLLY_ALWAYS_INLINE bool call(
       out_type<Varchar>& result,
       const arg_type<Varchar>& url) {
+    auto validUrl = urlUnescapeCheck(url);
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:176]: nit: validUrl variable is not needed

`if(!urlUnescapeCheck(url))`

File: velox/functions/prestosql/URLFunctions.h [hunk 5]
```diff
@@ -201,7 +246,8 @@ struct UrlExtractHostFunction {
     if (matchAuthorityAndPath(
             match, authAndPathMatch, authorityMatch, hasAuthority) &&
         hasAuthority) {
-      result.setNoCopy(submatch(authorityMatch, 3));
+      auto host = submatch(authorityMatch, 3);
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:249]: nit: host variable is not needed

File: velox/functions/prestosql/URLFunctions.h [hunk 6]
```diff
@@ -244,29 +294,25 @@ struct UrlExtractPathFunction {
...
-      } else {
-        escapedPath = submatch(match, 2);
-      }
-      urlUnescape(result, escapedPath);
+    if (!urlUnescapeCheck(url)) {
+      result.setEmpty();
+      return false;
     }
+
+    StringView escapedPath;
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:311]: escapedPath variable is not needed

**Review by mbasmanova — COMMENTED** [review]
@kgpai Thank you for the fix. Would you update documentation to clarify semantics of these functions?

File: velox/functions/prestosql/URLFunctions.h [hunk 7]
```diff
@@ -29,11 +30,33 @@ FOLLY_ALWAYS_INLINE StringView submatch(const boost::cmatch& match, int idx) {
 
 template <typename TInString>
 bool parse(const TInString& rawUrl, boost::cmatch& match) {
+  /// This regex is taken from RFC - 3986.
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:33]: Nice comment. Thanks.

File: velox/functions/prestosql/URLFunctions.h [hunk 8]
```diff
@@ -84,6 +107,37 @@ FOLLY_ALWAYS_INLINE void urlEscape(TOutString& output, const TInString& input) {
   output.resize(outIndex);
 }
 
+/// Performs basic initial validation of the URI.
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:110]: nit: 'basic initial' -> pick one, either basic or initial, no need to have both

File: velox/functions/prestosql/URLFunctions.h [hunk 9]
```diff
@@ -84,6 +107,37 @@ FOLLY_ALWAYS_INLINE void urlEscape(TOutString& output, const TInString& input) {
...
+/// Checks if the URI contains ascii whitespaces or
```
Comment (mbasmanova) [inline velox/functions/prestosql/URLFunctions.h:111]: I wonder if this check is complete? Are all ASCII characters other than whitespace allowed?

**Review by mbasmanova — COMMENTED** [review]
@kgpai Looks good, but let's update documentation in https://facebookincubator.github.io/velox/functions/presto/url.html to clarify what constitutes a valid URL and what happens if URL is not valid. Let's include some examples to helps readers understand.

**Comment by mbasmanova** [comment]
@kgpai Would you update PR description with some examples of before and after to help readers understand the impact of this change?

File: velox/functions/prestosql/URLFunctions.h (hunk 9 above)
Comment (kgpai) [reply velox/functions/prestosql/URLFunctions.h:111]: Yes ascii characters other than whitespaces are allowed.

File: velox/docs/functions/presto/url.rst [hunk 10]
```diff
@@ -14,6 +14,36 @@ The URL extraction functions extract components from HTTP URLs (or any valid URI
...
+    scheme    = http
+    authority = www.ics.uci.edu
+    path      = /pub/ietf/uri/
+    query     = k1=v1
+    fragment  = Related
+
+
+Invalid URI's
+-------------
+Well formed URI's should not contain ascii whitespace and percent-encoded URI's should be followed by two hexadecimal
```
Comment (mbasmanova) [inline velox/docs/functions/presto/url.rst:32]: Add empty line.

ascii -> ASCII

Consider splitting this sentence into two for readability.

```
Well formed URI's should not contain ascii whitespace. Percent-encoded URI's should be followed by two hexadecimal ...

Is "Percent-encoded URI's" a well-known term? Consider adding a link to a definition in Wikipedia or some other good source.

File: velox/docs/functions/presto/url.rst [hunk 11]
```diff
@@ -14,6 +14,36 @@ The URL extraction functions extract components from HTTP URLs (or any valid URI
...
+digits following the percent character "%".
+
+.. code-block::
+
+    # URI with whitespace
+    SELECT url_extract_path('foo '); -- NULL (1 row)
```
Comment (mbasmanova) [inline velox/docs/functions/presto/url.rst:38]: What do you want to show with this example? Are you trying to say that all URL functions return NULL when input URL is invalid? If so, please, state this directly.

**Review by mbasmanova — COMMENTED** [review]
Thank you for extending the docs. Some comments.

File: velox/docs/functions/presto/url.rst (hunk 10 above)
Comment (kgpai) [reply velox/docs/functions/presto/url.rst:32]: Will do. "Percent-encoded" is how its referred to in the URI RFC, I will reference that.

**Pull request closed without merging**